Automattic/newspack-newsletters
# feat: enable public newsletter posts

### All Submissions:






### Changes proposed in this Pull Request:



Allows editors to mark sent/published newsletters as publicly viewable and creates an archive page for newsletter posts. If a published newsletter is publicly viewable, its single page is visible to all users. If not, it's visible as a preview only to logged-in WP admin users (drafts are also visible to logged-in admin users). Also hijacks the main query on the Newsletter archive pages and filters out any posts that are NOT marked as publicly viewable.

#### Further work required?

* ~[Posts Inserter block isn't rendering anything on the front-end](https://github.com/Automattic/newspack-newsletters/pull/272#issuecomment-673016984)—need to fix that before we merge~ fixed in 91b596c

* Currently, newsletter ads are not rendered in the page view. I'm not sure this is necessary, but feel free to change my mind.

* There might be some further design/front-end work needed on the theme side if we want the archive and single templates for newsletter posts to differ from regular articles (there is probably no need for a byline, or the sidebar widgets, for example). (cc @thomasguillot and/or @laurelfulford)

* Maybe disable some interactions with other plugins—e.g. Jetpack Related Posts (this example done in this PR)

Closes #171.
Closes #214.

### How to test the changes in this Pull Request:

1. Check out/build this branch, and flush permalinks. Go to **Newsletters > All Newsletters** in WP admin.

2. For any newsletters that have been sent, observe a new "Preview" link when hovering over the item in the post list:

<img width="321" alt="Screen Shot 2020-08-12 at 10 05 55 AM" src="https://user-images.githubusercontent.com/2230142/90039140-7cbfea00-dc83-11ea-9aea-6769e4b9e8b7.png">

3. Click the Preview link. Observe that the newsletter is viewable as a single post page while logged into WP admin.

<img width="1248" alt="Screen Shot 2020-08-12 at 10 54 14 AM" src="https://user-images.githubusercontent.com/2230142/90043990-328e3700-dc8a-11ea-863c-0b00386023af.png">

4. Open a new Incognito window and attempt to visit the same newsletter post. Observe that the page is a 404 while not logged in.

<img width="1276" alt="Screen Shot 2020-08-12 at 10 57 58 AM" src="https://user-images.githubusercontent.com/2230142/90044317-b9431400-dc8a-11ea-962a-5ffab6961286.png">

5. Go to https://[YOUR SITE URL]/newsletter/ to view the newsletter archive page. Right now it should be empty while both logged in and not logged in (only posts explicitly marked as publicly viewable should appear on the archive page).

6. Edit the newsletter post. Observe a new toggle option in the Document settings sidebar, disabled by default:

<img width="254" alt="Screen Shot 2020-08-12 at 10 07 43 AM" src="https://user-images.githubusercontent.com/2230142/90039330-b7298700-dc83-11ea-99ff-5633312f4e15.png">

7. Enable the setting. Observe that the previously disabled "Sent" button is now enabled and says "Update." Click "Update" to save the new setting.

8. Go back to the "All Newsletters" post list. Hover over the post and observe that there's now a "View" link instead of "Preview." Also observe that the "Sent" notice should now have an additional "Published as a page" note attached.

<img width="434" alt="Screen Shot 2020-08-12 at 11 13 08 AM" src="https://user-images.githubusercontent.com/2230142/90045815-d547b500-dc8c-11ea-8a06-2b0c7380210c.png">

9. Refresh the archive page—now observe that your public post is listed while both logged in and not logged in.

10. Visit the newsletter post's single page. Observe that the public post is now visible while both logged in and not logged in.

11. Set up at least one campaign in the Newspack Campaigns plugin (can be either overlay or inline). Observe that the campaign will never render in the newsletter post's single page under any circumstances.

12. Edit the newsletter post again and disable the "Make newsletter page public?" setting, then save. Observe that the page is no longer visible in the archive or as a single page to non-logged-in users.

13. As a sanity check, create a new newsletter but DON'T send it. DO enable the "Make newsletter page public?" setting.
  - Observe that the "Send" button says "Send and Publish" when the setting is enabled.
  - Click "Save Draft," then confirm that the draft is visible to logged-in users only.

14. As another sanity check, ensure that sending newsletters with both Mailchimp and Constant Contact still works as before.

### Other information:
